Emailfrom

nstefanski opened the pull request:
Adds an option to specify a user to use as the sender of the notification. This can be any user in the course with the capability to add a Reengagement activity, or it can be set to a "Default Teacher" if a course template is being made for multiple teachers to use. The emailfrom user's name, email, and id can also be used as template variables. If a default teacher can't be found in the course at the time the message is sent, or if a specific user selected has had the capability removed, then messages will still default to sending from the support user.

Also adds an option to send notifications as an instant message instead of an email. Both the emailfrom and the instantmessage option can be used together, or one of the other can be used independently.

This change adds new database fields so it requires a version update.

tzerafnx committed:
add emailfrom option to form

tzerafnx committed:
add emailfrom field in db

tzerafnx committed:
helper function to get emailfrom user when sending message

tzerafnx committed:
add emailfrom user info to template variables

tzerafnx committed:
updated language to better match form

tzerafnx committed:
added notification type option to form

tzerafnx committed:
added instantmessage field to db

tzerafnx committed:
cleaned up extra return

tzerafnx committed:
update reengagement_send_notification with instant message option

tzerafnx committed:
style fixes for Travis CI

tzerafnx committed:
style fix for Travis CI

nstefanski commented:
@danmarsden I fixed all of the style issues in the PHP, I'm not really sure what's causing the remaining error with postgres, "version 9.5 is required and you are running 9.4.21", if you can offer any guidance I can make adjustments in my pull request. I can say that the plugin is working with our postgres db, but we're running version 11.5 currently.

tzerafnx committed:
use postgres 9.5 with Travis CI

nstefanski committed:
Merge branch 'instantmessage' into emailfrom

tzerafnx committed:
style fixes for Travis CI

tzerafnx committed:
added help string

nstefanski commented:
@danmarsden Dan you seem like the most active maintainer of this plugin so I thought I'd flag you here. Let me know if you're interested in making this part of the Reengagement plugin.

danmarsden commented on lib.php:
```diff
@@ -980,3 +999,31 @@ function reengagement_checkstart($course, $cm, $reengagement) {
+        $context = context_course::instance($reengagement->course);
+        if ($userid == 1) { // Default teacher, get first teacher in course.
+            global $DB;
+            $params = array('roleid' => 3, 'contextid' => $context->id);
```
hard-coded roleid - this should be a site-level setting for re-engagement that allows the admin to select a role that is used for teaching staff or better still, use a has_capability check instead.

danmarsden replied:
in fact - it would be better to use the get_course_contacts() style handling here with the $CFG->coursecontact settings instead of hard-coding this to a specific role.

nstefanski replied:
changed this to get_enrolled_users() with a permissions check for 'mod/reengagement:addinstance'

danmarsden commented on lib.php:
```diff
@@ -980,3 +999,31 @@ function reengagement_checkstart($course, $cm, $reengagement) {
+    $userid = $reengagement->emailfrom;
+    if ($userid > 0) {
+        $context = context_course::instance($reengagement->course);
+        if ($userid == 1) { // Default teacher, get first teacher in course.
```
better to use something like is siteadmin() style code - userid ===1 might not always be the siteadmin.

nstefanski replied:
This should be the variable REENGAGEMENT_EMAILFROM_TEACHER, but I take your point that the user with id 1 may not be the siteadmin, so I'm changing the value for default teacher to -1, 0 is still the support user, and then any integer above 0 will be the actual user.

danmarsden commented on lang/en/reengagement.php:
```diff
@@ -60,6 +60,10 @@
 $string['emaildelay_help'] = 'When module is set to notify users "after delay", this setting controls how long the delay is.';
+$string['emailfrom'] = 'Send notification from';
+$string['emailfrom_help'] = 'The user to send the message from, pending permissions check. Default teacher is the first teacher in the course, will send from support user if no teachers are found.';
+$string['emailfromsupport'] = 'Support User';
```
Moodle uses sentance case. 

Support User -> Support user.
Default Teacher -> Default teacher

nstefanski replied:
changed

danmarsden commented on db/install.xml:
```diff
@@ -27,6 +27,8 @@
         <FIELD NAME="remindercount" TYPE="int" LENGTH="3" NOTNULL="true" DEFAULT="1" SEQUENCE="false" COMMENT="how many times to send the email"/>
         <FIELD NAME="suppresstarget" TYPE="int" LENGTH="10" NOTNULL="true" DEFAULT="0" SEQUENCE="false" COMMENT="ID of module that, if completed, should prevent an email being sent out."/>
         <FIELD NAME="emaildelay" TYPE="int" LENGTH="10" NOTNULL="true" DEFAULT="604800" SEQUENCE="false" COMMENT="Time setting related to email"/>
+        <FIELD NAME="emailfrom" TYPE="int" LENGTH="10" NOTNULL="true" DEFAULT="0" SEQUENCE="false" COMMENT="ID of user to use as sender, 0 for support user, 1 for default teacher."/>
```
it makes me nervous to see xmldb files changed that don't include a version bump, it makes me wonder if this is hand-rolled or if you have used xmldb editor - can you please make sure you have used xmldb editor to add these fields to the install.xml file?

nstefanski replied:
Hi Dan,

I used XMLDB editor but my dev site is not connected to the github repo, so I had to copy and paste the updated file from there to here. I don't have a lot of experience with the XMLDB editor so I'm not sure how to do it in a better way, hope that clears things up?

danmarsden commented:
looks like this has sat here for quite a while! - sorry for the delay in getting back to this, but I've just posted a few comments in-line - in general this seems like a good idea, just the implementation with hard-coded roleids that would need to be fixed before we could merge this.

nstefanski committed:
REENGAGEMENT_EMAILFROM_TEACHER definition

changed value to -1 and used in reengagement_get_emailfrom function

nstefanski committed:
use sentence case

nstefanski committed:
changed comment

using -1 for default teacher since userid 1 could in theory be a valid teacher

nstefanski committed:
Update version.php

nstefanski committed:
change version for emailfrom and instantmessage

nstefanski committed:
add fields to sql query

nstefanski committed:
get teachers by capability

instead of looking for the specific teacher role, look for all users enrolled in the course with the capability to add a reengagement activity

nstefanski committed:
Merge pull request #2 from nstefanski/emailfrom-patch-1

patching several issues

nstefanski commented:
@danmarsden it took me a while to circle back to this, but I think I've addressed all of the issues with the original pull, let me know what you think
